src/sign_msg_auth_token.c

```diff
@@ -0,0 +1,215 @@
+        return false;
+    }
+
+    BEGIN_TRY {
+        TRY {
+            cx_eddsa_sign(&private_key, CX_RND_RFC6979 | CX_LAST, CX_SHA512, token_auth_context.hash, HASH_LEN, NULL, 0, token_auth_context.signature, MESSAGE_SIGNATURE_LEN, NULL);
+        }
+        CATCH_ALL {
+            success = false;
+        }
+        FINALLY {
+            memset(&private_key, 0, sizeof(private_key));
+        }
+    }
+    END_TRY;
+
+    return success;
+}
+
+void handle_auth_token(uint8_t p1, uint8_t *data_buffer, uint16_t data_length, volatile unsigned int *flags) {
+    /* 
+        data buffer structure should be:
+        <account index> + <address index> + <token length> + <token>
+               ^                 ^                 ^            ^
+           4 bytes           4 bytes           4 bytes     <token length> bytes
+
+        the account and address indexes, alongside token length are computed in the first bulk, while the entire token can come in multiple bulks
+    */
+    if (p1 == P1_FIRST) {
+        memset(token_auth_context.token, 0, sizeof(token_auth_context.token));
+        token_auth_context.token[0] = '\0';
+        char token_length_str[11];
+
+        // check that the indexes and the length are valid
+        if (data_length < 12) {
+            THROW(ERR_INVALID_MESSAGE);
+        }
+
+        uint8_t public_key[PUBLIC_KEY_LEN];
+
+        uint32_t const account_index = read_uint32_be(data_buffer);
+        uint32_t const address_index = read_uint32_be(data_buffer + sizeof(uint32_t));
+        if (!get_public_key(account_index, address_index, public_key)) {
+            THROW(ERR_INVALID_ARGUMENTS);
+        }
+
+        get_address_bech32_from_binary(public_key, token_auth_context.address);
+
+        app_state = APP_STATE_SIGNING_MESSAGE;
+    
+        // account and address indexes (4 bytes each) have been read, so skip the first 8 bytes
+        data_buffer += 8;
+        data_length -= 8;
+ 
+        token_auth_context.len = U4BE(data_buffer, 0);
+
+        // the token length (4 bytes) has been read, so skip the next 4 bytes
+        data_buffer += 4;
+        data_length -= 4;
+
+        update_token_display_data(data_buffer, data_length);
+
+        // initialize hash with the constant string to prepend
+        cx_keccak_init(&sha3_context, SHA3_KECCAK_BITS);
+        cx_hash((cx_hash_t *)&sha3_context, 0, (uint8_t*)PREPEND, sizeof(PREPEND) - 1, NULL, 0);
+
+        // convert message length to string and store it in the variable `tmp`
+        uint32_t full_message_len = token_auth_context.len + BECH32_ADDRESS_LEN;
+        uint32_t_to_char_array(full_message_len, token_length_str);
+
+        // add the message length to the hash
+        cx_hash((cx_hash_t *)&sha3_context, 0, (uint8_t*)token_length_str, strlen(token_length_str), NULL, 0);
+
+        // add the message length to the hash
+        cx_hash((cx_hash_t *)&sha3_context, 0, token_auth_context.address, strlen(token_auth_context.address), NULL, 0);
```

Comment:
Cast `token_auth_context.address` to `uint8_t *` to avoid a compilation warning.

Reply:
done